### PR TITLE
chemberta in deepchem

### PR DESCRIPTION
**Pull request by arunppsg**
## Description

Adding support for chemberta model in deepchem

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/chemberta.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+class Chemberta(HuggingFaceModel):
+
+    def __init__(self,
+                 task: str,
+                 mode: Optional[str] = None,
+                 tokenizer_path: str = 'seyonec/PubChem10M_SMILES_BPE_60k',
+                 **kwargs):
+        tokenizer = RobertaTokenizerFast.from_pretrained(tokenizer_path)
+        if task == 'pretraining':
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:17]: There should be two pretraining tasks, mlm-pretraining, mtr-pretraining following chemberta-2 paper

File: deepchem/models/torch_models/chemberta.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    >>> finetuning_loss = finetune_model.fit(dataset, nb_epoch=1)
+
+    >>> # prediction and evaluation
+    >>> result = finetune_model.predict(dataset)
+    >>> eval_results = finetune_model.evaluate(dataset, metrics=dc.metrics.Metric(dc.metrics.mae_score))
+
+
+    Reference
+    ---------
+    .. Ahmad, Walid, et al. "Chemberta-2: Towards chemical foundation models." arXiv preprint arXiv:2209.01712 (2022).
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:74]: Can you add the chemberta-1 paper here?

File: deepchem/models/torch_models/chemberta.py (hunk 1 above)
Comment (rbharath) [reply deepchem/models/torch_models/chemberta.py:87]: Documenting offline discussion, we should follow same convention @tonydavis629 is following in the GCNModular for now, but right fix is to implement the Task API in subsequent PRs

File: deepchem/models/torch_models/chemberta.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                raise ValueError('invalid mode specification')
+        else:
+            raise ValueError('invalid task specification')
+
+        super(Chemberta, self).__init__(model=model,
+                                        task=task,
+                                        tokenizer=tokenizer,
+                                        **kwargs)
+
+    def _prepare_batch(self, batch: Tuple[Any, Any, Any]):
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:116]: Ok to move this to HFModel in a separate PR, but we should eventually create a BatchPreparer API to make this logic part of the public API

File: deepchem/models/torch_models/hf_models.py [hunk 4]
```diff
@@ -373,6 +373,10 @@ def _predict(self, generator: Iterable[Tuple[Any, Any, Any]],
             a NumPy array of the model produces a single output, or a list of arrays
             if it produces multiple outputs
         """
+        # FIXME A HuggingFace model does not output uncertainity. The variable is here
```
Comment (rbharath) [inline deepchem/models/torch_models/hf_models.py:376]: We should add this to the docstring

File: deepchem/models/torch_models/tests/test_chemberta.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    model_new.restore()
+
+    old_state = model.model.state_dict()
+    new_state = model_new.model.state_dict()
+    matches = [
+        torch.allclose(old_state[key], new_state[key])
+        for key in old_state.keys()
+    ]
+
+    # all keys values should match
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_chemberta.py:124]: Can we add a unit test checking that we can load the public chemberta weights (assuming this test isn't too slow)

**Review by rbharath — COMMENTED** [review]
Looking mostly ready, mostly minor comments below. @arunppsg to address batch preparation for HFModel in separate PR and return to this one.

File: deepchem/models/torch_models/chemberta.py (hunk 1 above)
Comment (arunppsg) [reply deepchem/models/torch_models/chemberta.py:87]: If I am not wrong, GNNModular uses an attribute named `task`to denote the type of unsupervised task. Am I correct here @tonydavis629 ? 

I will go ahead and use the same.

File: deepchem/models/torch_models/chemberta.py (hunk 2 above)
Comment (arunppsg) [reply deepchem/models/torch_models/chemberta.py:74]: added chemberta-1 paper

File: deepchem/models/torch_models/chemberta.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+
+    Reference
+    ---------
+    .. Ahmad, Walid, et al. "Chemberta-2: Towards chemical foundation models." arXiv preprint arXiv:2209.01712 (2022).
+    """
+
+    def __init__(self,
+                 task: str,
+                 mode: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (arunppsg) [reply deepchem/models/torch_models/chemberta.py:79]: added n_tasks

File: deepchem/models/torch_models/chemberta.py (hunk 1 above)
Comment (arunppsg) [reply deepchem/models/torch_models/chemberta.py:87]: added task here as well as updated in HuggingFaceModel base class in #3389

File: deepchem/models/torch_models/chemberta.py (hunk 3 above)
Comment (arunppsg) [reply deepchem/models/torch_models/chemberta.py:116]: Moved to separate PR #3389

File: deepchem/models/torch_models/hf_models.py (hunk 4 above)
Comment (arunppsg) [reply deepchem/models/torch_models/hf_models.py:376]: added to docstring in a separate PR

File: deepchem/models/torch_models/chemberta.py (hunk 1 above)
Comment (arunppsg) [reply deepchem/models/torch_models/chemberta.py]: makes sense, I have updated to remove the use of `mode` and to use only `task`.

File: deepchem/models/torch_models/tests/test_chemberta.py (hunk 5 above)
Comment (arunppsg) [reply deepchem/models/torch_models/tests/test_chemberta.py:124]: added test in #3390. Since this class reuses implementation from base model, I think a test in base class will be sufficient.

File: docs/source/api_reference/models.rst [hunk 7]
```diff
@@ -490,35 +482,49 @@ Density Functional Theory Model - XCModel
...
+
+Hugging Face Models
+===================
+
+HuggingFace models from the `transformers <https://huggingface.co/models>`_ library can wrapped using the wrapper :code:`HuggingFaceModel`
+
+.. autoclass:: deepchem.models.torch_models.hf_models.HuggingFaceModel
+  :members:
+
+Chemberta
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:526]: Can you add this to the model CSV for the cheatsheet?

**Review by rbharath — APPROVED** [review]
LGTM, a couple of minor comments to address in the next follow-up PR

**Comment by Zrahay** [comment]
Hi @arunppsg! I am a current GSoC mentee with DeepChem. I am working on implementing DNABERT-2 in DeepChem and I am closely following Chemberta.py for my implementation. I saw this file was committed to you so I wanted to ask if I could get your Discord username so that I could ask you some questions which might help me in my implementation.

Thanks!
